File: frontend/src/types/acquisition.ts
import type { ApiResponse, EntityType } from '@/types/metadata';
import type {
  MusicMediaInfo,
  MusicMediaInput,
  MusicMetaBase,
  MusicRecognitionAssessment,
} from '@/types/music-media';

export type TriggerSource = 'manual' | 'chart' | 'subscription' | 'artist_watch';
export type JobStatus =
  | 'queued'
  | 'running'
  | 'matched'
  | 'manual_pending'
  | 'dispatched'
  | 'completed'
  | 'no_result'
  | 'failed';
export type DecisionStatus = 'auto_download' | 'manual_confirm' | 'reject' | 'pending';
export type AdapterMode = 'mock' | 'host';
export type AdapterSelectionMode = 'mock' | 'prefer_host' | 'strict_host';
export type VerificationState = 'verified' | 'unverified' | 'placeholder';

export interface AdapterResolution {
  adapter_key: string;
  adapter_mode: AdapterMode;
  selection_mode: AdapterSelectionMode;
  capability_source: string;
  verification_state: VerificationState;
  fallback_reason?: string | null;
  integration_point: string;
  host_integration_enabled: boolean;
}

export interface PathHandoffInfo {
  download_hash?: string | null;
  source_path?: string | null;
  source_filetype?: string | null;
  source_name?: string | null;
  source_basename?: string | null;
  source_extension?: string | null;
  handoff_source: string;
  handoff_status: string;
  verification_state: VerificationState;
  note: string;
  raw_summary: Record<string, unknown>;
}

export interface QueryPreferences {
  preferred_formats: string[];
  prefer_lossless: boolean;
  include_aliases: boolean;
  include_year: boolean;
  allow_live: boolean;
  allow_remaster: boolean;
  negative_keywords: string[];
  auto_download_threshold: number;
  manual_confirm_threshold: number;
}

export interface QueryClause {
  query_type: 'canonical' | 'alias' | 'relaxed' | 'negative';
  source: string;
  query: string;
  explanation: string;
  priority: number;
}

export interface QueryContext {
  entity_type: EntityType;
  provider: string;
  provider_id: string;
  title: string;
  artist_names: string[];
  album_title?: string | null;
  album_artist_names: string[];
  year?: number | null;
  track_number?: number | null;
  disc_number?: number | null;
  external_refs: Record<string, string>;
  match_strategy?: string | null;
  note: string;
  summary: string;
}

export interface QueryBuildResult {
  entity_type: EntityType;
  provider: string;
  provider_id: string;
  music_media_info: MusicMediaInfo;
  mock: boolean;
  preferences: QueryPreferences;
  canonical_queries: QueryClause[];
  alias_queries: QueryClause[];
  relaxed_queries: QueryClause[];
  negative_queries: QueryClause[];
  ordered_queries: QueryClause[];
  query_context: QueryContext;
  note: string;
  integration_point: string;
  todo: string[];
}

export interface SearchJobSummary {
  id: string;
  music_media_input: MusicMediaInput;
  music_meta_base: MusicMetaBase;
  music_recognition_assessment: MusicRecognitionAssessment;
  music_media_info: MusicMediaInfo;
  trigger_source: TriggerSource;
  profile_id: string;
  mode: string;
  status: JobStatus;
  created_at: string;
  updated_at: string;
  started_at?: string | null;
  finished_at?: string | null;
  mock: boolean;
  note?: string | null;
  query_build?: QueryBuildResult | null;
  summary: Record<string, unknown>;
  error_message?: string | null;
  adapter_resolution?: AdapterResolution | null;
}

export interface ScoreBreakdownItem {
  score: number;
  reason: string;
}

export interface SearchCandidateDetail {
  id: string;
  job_id: string;
  site_id: string;
  site_name: string;
  title: string;
  normalized_title: string;
  size_bytes: number;
  seeders: number;
  peers: number;
  format_tag?: string | null;
  bitrate_kbps?: number | null;
  source_tags: string[];
  raw_score: number;
  score_total: number;
  score_breakdown: Record<string, ScoreBreakdownItem>;
  decision: DecisionStatus;
  reason_codes: string[];
  dispatchable: boolean;
  dispatch_status: string;
  mock: boolean;
  note?: string | null;
  created_at: string;
  adapter_resolution?: AdapterResolution | null;
  path_handoff?: PathHandoffInfo | null;
  raw_payload?: Record<string, unknown>;
}

export interface SearchCandidateListData {
  job_id: string;
  items: SearchCandidateDetail[];
  total: number;
  mock: boolean;
  note: string;
  adapter_resolution?: AdapterResolution | null;
}

export interface SearchJobCreatePayload {
  input: MusicMediaInput;
  trigger_source?: TriggerSource;
  profile_id?: string;
  mode?: 'manual' | 'auto';
  preferences?: Partial<QueryPreferences>;
}

export interface QueryBuildPayload {
  input: MusicMediaInput;
  preferences?: Partial<QueryPreferences>;
}

export interface DispatchPayload {
  result_id: string;
  downloader_id?: string;
  save_path_policy?: 'auto' | 'manual';
  manual_confirm?: boolean;
}

export interface DispatchResult {
  candidate_id: string;
  job_id: string;
  dispatchable: boolean;
  dispatch_status: string;
  target_downloader: string;
  downloader_task_id?: string | null;
  note: string;
  integration_point: string;
  mock: boolean;
  binding_id?: string | null;
  dispatch_backend: AdapterMode;
  capability_source: string;
  fallback_reason?: string | null;
  failure_reason?: string | null;
  verification_state: VerificationState;
  path_handoff?: PathHandoffInfo | null;
  host_response_summary: Record<string, unknown>;
  adapter_resolution?: AdapterResolution | null;
}

export type ApiQueryBuildResponse = ApiResponse<QueryBuildResult>;
export type ApiSearchJobListResponse = ApiResponse<SearchJobSummary[]>;
export type ApiSearchJobResponse = ApiResponse<SearchJobSummary>;
export type ApiSearchCandidatesResponse = ApiResponse<SearchCandidateListData>;
export type ApiDispatchResponse = ApiResponse<DispatchResult>;

File: backend/app/schemas/acquisition.py
"""Schemas for Phase 3 query building, search jobs, scoring, and dispatch."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

from .integration import AdapterMode, AdapterResolution, VerificationState
from .music_media import (
    MusicMediaInfo,
    MusicMediaInput,
    MusicMediaMatchStrategy,
    MusicMetaBase,
    MusicRecognitionAssessment,
)
from .shared import DecisionStatus, EntityType, JobStatus, TriggerSource


class QueryPreferences(BaseModel):
    preferred_formats: list[str] = Field(default_factory=lambda: ["FLAC"])
    prefer_lossless: bool = True
    include_aliases: bool = True
    include_year: bool = True
    allow_live: bool = False
    allow_remaster: bool = False
    negative_keywords: list[str] = Field(default_factory=list)
    auto_download_threshold: float = Field(default=90.0, ge=0, le=100)
    manual_confirm_threshold: float = Field(default=70.0, ge=0, le=100)


class QueryBuildRequest(BaseModel):
    input: MusicMediaInput
    preferences: QueryPreferences = Field(default_factory=QueryPreferences)


class QueryClause(BaseModel):
    query_type: Literal["canonical", "alias", "relaxed", "negative"]
    source: str
    query: str
    explanation: str
    priority: int = Field(default=100, ge=1)


class QueryContext(BaseModel):
    entity_type: EntityType
    provider: str
    provider_id: str
    title: str
    artist_names: list[str] = Field(default_factory=list)
    album_title: str | None = None
    album_artist_names: list[str] = Field(default_factory=list)
    year: int | None = None
    track_number: int | None = None
    disc_number: int | None = None
    external_refs: dict[str, str] = Field(default_factory=dict)
    match_strategy: MusicMediaMatchStrategy | None = None
    note: str
    summary: str


class QueryBuildResult(BaseModel):
    entity_type: EntityType
    provider: str
    provider_id: str
    music_media_info: MusicMediaInfo
    mock: bool = True
    preferences: QueryPreferences
    canonical_queries: list[QueryClause] = Field(default_factory=list)
    alias_queries: list[QueryClause] = Field(default_factory=list)
    relaxed_queries: list[QueryClause] = Field(default_factory=list)
    negative_queries: list[QueryClause] = Field(default_factory=list)
    ordered_queries: list[QueryClause] = Field(default_factory=list)
    query_context: QueryContext
    note: str
    integration_point: str
    todo: list[str] = Field(default_factory=list)


class HostSearchCandidate(BaseModel):
    site_id: str
    site_name: str
    title: str
    normalized_title: str
    size_bytes: int = 0
    seeders: int = 0
    peers: int = 0
    format_tag: str | None = None
    bitrate_kbps: int | None = None
    source_tags: list[str] = Field(default_factory=list)
    mock: bool = True
    note: str
    adapter_resolution: AdapterResolution | None = None
    raw_payload: dict[str, Any] = Field(default_factory=dict)


class PathHandoffInfo(BaseModel):
    download_hash: str | None = None
    source_path: str | None = None
    source_filetype: str | None = None
    source_name: str | None = None
    source_basename: str | None = None
    source_extension: str | None = None
    handoff_source: str
    handoff_status: str
    verification_state: VerificationState = VerificationState.UNVERIFIED
    note: str
    raw_summary: dict[str, Any] = Field(default_factory=dict)


class ScoreBreakdownItem(BaseModel):
    score: float
    reason: str


class CandidateScoreResult(BaseModel):
    raw_score: float
    score_total: float
    score_breakdown: dict[str, ScoreBreakdownItem]
    decision: DecisionStatus
    reason_codes: list[str] = Field(default_factory=list)
    dispatchable: bool = False


class SearchJobCreateRequest(BaseModel):
    input: MusicMediaInput
    trigger_source: TriggerSource = TriggerSource.MANUAL
    profile_id: str = "default-lossless"
    mode: Literal["manual", "auto"] = "manual"
    preferences: QueryPreferences = Field(default_factory=QueryPreferences)


class SearchJobSummary(BaseModel):
    id: str
    music_media_input: MusicMediaInput
    music_meta_base: MusicMetaBase
    music_recognition_assessment: MusicRecognitionAssessment
    music_media_info: MusicMediaInfo
    trigger_source: TriggerSource
    profile_id: str
    mode: str
    status: JobStatus
    created_at: datetime
    updated_at: datetime
    started_at: datetime | None = None
    finished_at: datetime | None = None
    mock: bool = True
    note: str | None = None
    query_build: QueryBuildResult | None = None
    summary: dict[str, Any] = Field(default_factory=dict)
    error_message: str | None = None
    adapter_resolution: AdapterResolution | None = None


class SearchCandidateDetail(BaseModel):
    id: str
    job_id: str
    site_id: str
    site_name: str
    title: str
    normalized_title: str
    size_bytes: int
    seeders: int
    peers: int
    format_tag: str | None = None
    bitrate_kbps: int | None = None
    source_tags: list[str] = Field(default_factory=list)
    raw_score: float
    score_total: float
    score_breakdown: dict[str, ScoreBreakdownItem] = Field(default_factory=dict)
    decision: DecisionStatus
    reason_codes: list[str] = Field(default_factory=list)
    dispatchable: bool = False
    dispatch_status: str = "pending"
    mock: bool = True
    note: str | None = None
    created_at: datetime
    adapter_resolution: AdapterResolution | None = None
    path_handoff: PathHandoffInfo | None = None
    raw_payload: dict[str, Any] = Field(default_factory=dict)


class SearchCandidateListData(BaseModel):
    job_id: str
    items: list[SearchCandidateDetail] = Field(default_factory=list)
    total: int = 0
    mock: bool = True
    note: str
    adapter_resolution: AdapterResolution | None = None


class DownloadBindingSummary(BaseModel):
    id: str
    job_id: str
    candidate_id: str
    target_downloader: str
    downloader_task_id: str | None = None
    dispatchable: bool = False
    dispatch_status: str
    mock: bool = True
    note: str | None = None
    integration_point: str | None = None
    dispatched_at: datetime
    path_handoff: PathHandoffInfo | None = None
    host_response_summary: dict[str, Any] = Field(default_factory=dict)


class DownloadBindingDetail(DownloadBindingSummary):
    candidate: SearchCandidateDetail | None = None
    raw_payload: dict[str, Any] = Field(default_factory=dict)


class DownloadBindingListData(BaseModel):
    items: list[DownloadBindingSummary] = Field(default_factory=list)
    total: int = 0
    mock: bool = True
    note: str


class MutationResult(BaseModel):
    id: str
    deleted: bool = False


class DispatchRequest(BaseModel):
    result_id: str
    downloader_id: str = "mock-downloader"
    save_path_policy: Literal["auto", "manual"] = "auto"
    manual_confirm: bool = True


class DispatchResult(BaseModel):
    candidate_id: str
    job_id: str
    dispatchable: bool
    dispatch_status: str
    target_downloader: str
    downloader_task_id: str | None = None
    note: str
    integration_point: str
    mock: bool = True
    binding_id: str | None = None
    dispatch_backend: AdapterMode = AdapterMode.MOCK
    capability_source: str = "mock.adapter"
    fallback_reason: str | None = None
    failure_reason: str | None = None
    verification_state: VerificationState = VerificationState.PLACEHOLDER
    path_handoff: PathHandoffInfo | None = None
    host_response_summary: dict[str, Any] = Field(default_factory=dict)
    adapter_resolution: AdapterResolution | None = None


class DispatchAdapterResult(BaseModel):
    dispatchable: bool
    dispatch_status: str
    target_downloader: str
    downloader_task_id: str | None = None
    note: str
    integration_point: str
    mock: bool = True
    dispatch_backend: AdapterMode = AdapterMode.MOCK
    capability_source: str = "mock.adapter"
    fallback_reason: str | None = None
    failure_reason: str | None = None
    verification_state: VerificationState = VerificationState.PLACEHOLDER
    path_handoff: PathHandoffInfo | None = None
    host_response_summary: dict[str, Any] = Field(default_factory=dict)
    adapter_resolution: AdapterResolution | None = None
